pkg/tsdb/tempo/grpc.go

```diff
@@ -74,6 +76,39 @@ func getDialOpts(settings backend.DataSourceInstanceSettings, opts httpclient.Op
 		dialOps = append(dialOps, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	}
 
+	// The following code is required to make gRPC work with Grafana Cloud PDC
+	// (https://grafana.com/docs/grafana-cloud/connect-externally-hosted/private-data-source-connect/)
+	proxyClient, err := settings.ProxyClient(ctx)
+	if err != nil {
+		logger.Error("Error getting proxy client. The Tempo plugin will not use the secure socks proxy", "error", err)
+		return dialOps
+	}
+	if proxyClient.SecureSocksProxyEnabled() { // secure socks proxy is behind a feature flag
+		dialer, err := proxyClient.NewSecureSocksProxyContextDialer()
+		if err != nil {
+			logger.Error("Error dialing secure socks proxy. The Tempo plugin will not use the secure socks proxy", "error", err)
+			return dialOps
+		}
+		logger.Debug("gRPC dialer instantiated. Appending gRPC dialer to dial options")
+		dialOps = append(dialOps, grpc.WithContextDialer(func(ctx context.Context, host string) (net.Conn, error) {
+			select {
+			case <-ctx.Done():
+				logger.Warn("Context has been canceled, this should not happen. Proceeding with dialing anyway")
+			default:
+				logger.Debug("Context is still valid, proceeding with dialing")
+			}
+
+			logger.Debug("Dialing secure socks proxy", "host", host)
+			conn, err := dialer.Dial("tcp", host)
+			if err != nil {
+				logger.Error("Error dialing secure socks proxy. Returning connection anyway", "error", err)
```

Comment:
Again what will happen if we have error and return the connection anyway? Seems like that would propagate invalid connection up, creating additional errors that could mask the root cause.